plugins/pulsar-records-storage/src/test/java/com/github/bsideup/liiklus/pulsar/PulsarRecordsStorageTest.java

```diff
@@ -70,4 +91,212 @@ public String keyByPartition(int partition) {
+    public void testExclusiveRecordDistribution() throws Exception {
+        var numberOfPartitions = getNumberOfPartitions();
+        Assumptions.assumeTrue(numberOfPartitions > 1, "target supports more than 1 partition");
+
+        var groupName = UUID.randomUUID().toString();
+
+        var receivedOffsets = new ConcurrentHashMap<RecordsStorage.Subscription, Set<Tuple2<Integer, Long>>>();
+
+        var disposeAll = ReplayProcessor.<Boolean>create(1);
+
+        Function<RecordsStorage.Subscription, Disposable> subscribeAndAssign = subscription -> {
+            return Flux.from(subscription.getPublisher(() -> CompletableFuture.completedFuture(Collections.emptyMap())))
+                    .flatMap(Flux::fromStream, numberOfPartitions)
+                    .flatMap(RecordsStorage.PartitionSource::getPublisher, numberOfPartitions)
+                    .takeUntilOther(disposeAll)
+                    .subscribe(record -> {
+                        receivedOffsets
+                                .computeIfAbsent(subscription, __ -> new HashSet<>())
+                                .add(Tuples.of(record.getPartition(), record.getOffset()));
+                    });
+        };
+
+        try {
+            var firstSubscription = getTarget().subscribe(getTopic(), groupName, Optional.of("earliest"));
+            var secondSubscription = getTarget().subscribe(getTopic(), groupName, Optional.of("earliest"));
+
+            subscribeAndAssign.apply(firstSubscription);
+            await.untilAsserted(() -> {
+                try {
+                    assertThat(receivedOffsets)
+                            .containsKeys(firstSubscription)
+                            .doesNotContainKey(secondSubscription)
+                            .allSatisfy((key, value) -> assertThat(value).isNotEmpty());
+                } catch (Throwable e) {
+                    publishToEveryPartition();
+                    throw e;
+                }
+            });
+
+            subscribeAndAssign.apply(secondSubscription);
+            publishToEveryPartition();
+
+            await.pollDelay(org.awaitility.Duration.TWO_SECONDS).untilAsserted(() -> {
+                try {
+                    assertThat(receivedOffsets)
+                            .containsKeys(firstSubscription)
+                            .doesNotContainKey(secondSubscription)
+                            .allSatisfy((key, value) -> assertThat(value).isNotEmpty());
+                } catch (Throwable e) {
+                    publishToEveryPartition();
+                    throw e;
+                }
+            });
+        } finally {
+            disposeAll.onNext(true);
+        }
+    }
+
+    @Test
+    @Override
+    // currently each liiklus consumers consume all partition, and leaves the client to rebalance
+    // in pulsar, the pulsar client does not rebalance naturally on failover or exclusive
+    public void testMultipleGroups() throws Exception {
```

Comment:
this test is pretty storage-agnostic, but does it have to be overridden? I am afraid we're introducing a regression here

Reply:
I am not sure it's the regression, if anything I would say that the tests depends too much on the intrinsic behaviour of the underlying storage. 

e.g. for kafka, we can have two consumers for a group name, and there will be rebalancing, and the rebalancing is working properly. e.g. if one consumer disconnnect, the second consumer will do a reset offset based on liiklus offset and continue consuming. However, in pulsar failover mode, having two consumers, doing an offset seek will disconnect all consumers, and cause the second consumer to also be stuck.

Reply:
Doesn't exclusive mode allow only one consumer to be connected at a time? So that we get assignment-like behaviour, assumed by this test

Reply:
it differs in subscription time, or let me check the tests in more detail

Reply:
**testExclusiveRecordDistribution**
here the test checks that if two consumers connect at the same time, that each consumer will have messages consumed. however this is not correct, as the rebalancing can happens that only one consumer is the one handling all partitions. this is not very liiklus specific, but dependant on the records storage native client implementation.

**testMultipleGroups**
here the test checks that if one consumer is connected, that it's getting all messages. however the next thing tested is that if a second consumer is connected. the second consumer also have consumed some messages. this is also the wrong assumption as above as liiklus does not do any balancing, but the records storage native client implementation does.


Checking above tests, since liiklus always assign all partitions to any connecting consumers, the rebalancing happens based on the records storage native client implementation. So, it is the test which is not consistent, or in other words, the test is expecting some native client implementation behaviour (in this case pulsar). 

or am I missing some understanding here?

Reply:
> the rebalancing can happens that only one consumer is the one handling all partitions

Both Kafka and Pulsar support eventual rebalancing. It is a very important aspect of such systems, and, if this test fails, we may have a problem because it means that one consumer will be holding a "lock" without giving others opportunity to be assigned.
Maybe this is why Kafka emulation in Pulsar is using `Failover`?
https://github.com/apache/pulsar/blob/83365ebd536af3b3575026a06d0b83cd506898d3/pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/consumer/PulsarKafkaConsumer.java#L251

Same about `testMultipleGroups`

After reviewing this, I have a feeling that `exclusive` may not be a good fit here. It does not support the rebalancing (unlike `failover`).

I wonder if the seek fix also fixes the rebalancing with `failover`?

Reply:
unfortunately, manual seek seems to cause the `failover` to stop working, at least in our tests the if consumer A and B connected to a subscription with failover, and A is getting messages. B will not continue if A is disconnected, which is why we switched to `exclusive`. Currently, this patch is deployed in our staging, and is the only variant which does not include occasional lags.

Reply:
It works, but without the rebalancing, I assume?

FYI I asked on Slack about it:
https://apache-pulsar.slack.com/archives/C5Z4T36F7/p1567090931044200

Reply:
it rebalance only if a consumer of a partition died because of any reason.

Reply:
well, it falls back, but the "rebalancing" is not happening (where "rebalancing" is re-assigning partitions between consumers to balance the load)

Reply:
Which makes me wonder.... unlike in Kafka, in Pulsar, partitions are topics, and there is N consumers for each "partitioned topic". Which means that Pulsar cannot detect that N consumers are coming from the same "process", hence no rebalancing.

Maybe there is some property to set when "exclusive" is used to make it detect overassigned instances?

Reply:
Pulsar's abstraction is a bit different than Kafka. In Kafka, there is a partition -> consumer. I don't think the partition abstraction is that different (except that in pulsar it's exposed as another topic internally). In pulsar, there is a partition -> subscription -> consumer.

So, a pulsar client does not really connect to a partition, but it connects to a subscription (based on a subscription name). In turn, subscription is connected to 1 topic or more (in our case it's only 1 subscription (per 1 subscription name) per 1 partition. Exclusive means, that only one consumer can be connected to a subscription. With Failover, it means, we can have multiple consumers connected to a subscription, and if a consumer die, pulsar will push all unacked messages to other consumer. So there is not really any rebalancing happening, as if the first consumer is connected, the message in the partition will go to that consumer in any subscription mode.

However, in liiklus, we do a manual seek with the consumer. The offset of a partition actually lies with the subscription. Hence if a manual seek is done. Pulsar disconnect all consumers of a subscription. Hence, a rebalancing might happen on reconnect as the one to reconnect first might be another consumer. However this is a side effect of the last pulsar records storage implementation. The next issue is that, because of this seek. For some reason, if the receiving consumer is disconnected, the other connected consumer won't get any message.

Additionally, this is not further investigated yet. Since the seek only happens on initial subscription. It means in case of a failover switch, the second consumer will get all messages since the last seek as liiklus does not ack any messages, and in case of failover pulsar send to the other consumer all un-acked messages (based on the last saved offset in the subscription). However, as mentioned above, this does not happens in the test since we only test new consumer connection, and did not really test what happens on failover, is the one receiving message is killed.

So in short, to my understanding pulsar does not really rebalance automatically and we can't assume it will happen on new consumer creation.

p.s. there is actually a way to have always a rebalancing, and that is the Key_Shared subscription mode. However, liiklus offset and ack does not really support this in a good way as in pulsar, ack is actually done on a message basis, and in liiklus it's done on an offset basis.

Reply:
>  Pulsar disconnect all consumers of a subscription.

FYI it is fixed in master: https://apache-pulsar.slack.com/archives/C5Z4T36F7/p1567098757051900?thread_ts=1567090931.044200&cid=C5Z4T36F7

(I will comment more, just wanted to point out eagerly)